### PR TITLE
extmod/framebuf: Replace *sb_*pixel functions with a single generic one

### PR DESCRIPTION
**Pull request by deshipu**
Use a single pixel function with parameters for different number of bits
per pixel, bit order and pixel organization. Allows adding more modes
without writing much extra code.

Note: we specify both bpp (bits per pixel) and log_ppb (log2 of pixels
per byte), because calculating one from the other would be expensive.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for this and sorry for the late response.  It's great to reduce code size, but the main thing here would be to assess the loss in performance (if any) and see if it is acceptable.  That would likely take some effort, to make tests and run them on various hardware.

**Comment by deshipu** [comment]
I agree it would be great, but to be honest I doubt anybody cares about this enough to expend the effort. I did this because you requested it in https://github.com/micropython/micropython/pull/3214 but now you are moving the goalposts. Personally, I doubt this patch will reduce the code size by much or affect speed, because if you look closely, the shared function is inlined -- hopefully the compiler would be smart enough to optimize out the constant parameters in each case, and generate code that is not really that dissimilar from what is there right now. So the gain is mostly in simplifying the code, making it easier to add more modes, and perhaps reducing the risk of bugs.

**Comment by dpgeorge** [comment]
The change in code size with this patch is:
```
      stm32:  -248
    esp8266:  -528
```
(Other ports don't use this module.)

**Comment by pfalcon** [comment]
-1 for the change of this patch. framebuf module is a rare case when performance should be considered seriously in uPy: there's a big difference e.g. with being able to do even a small, but animation, vs not being able to do, instead getting torn/flickering frame transitions. And having dedicated, optimized routines for different bit depth/sizes/etc. is a well known technique to achieve it. Some similar rarely used routines perhaps could be compressed into one, and I'm sure that's what @dpgeorge meant. But -1 for lumping everything together.

**Comment by deshipu** [comment]
Well, if that's out of the way then, how about merging https://github.com/micropython/micropython/pull/3214?

**Comment by dpgeorge** [comment]
@deshipu I appreciate you working on this and providing the PR, but it may end up that this demonstrates how not to do things.  The MicroPython code needs to balance a very thin (and sometimes impossible) line between performance, features and size.  Quite often I have worked for days/weeks on a new feature, or optimisation, only to find that it makes things worse and then I need to discard a lot of hard work.  It's like scientific research: sometimes you end up at a dead end, but at least it helps you to know where not to go.  That's just the nature of this project, and also what makes it fun (at least for me) to work on.

I would like to do performance tests with this PR to see exactly how things change.

**Comment by dpgeorge** [comment]
> Well, if that's out of the way then, how about merging #3214?

Optimisation was also mentioned in that PR, and it seems the submitter did not have time to look into that.

**Comment by peterhinch** [comment]
I thought it might be interesting to perform some actual measurements.
tl;dr There is a substantial performance penalty with 4-bit and 1-bit formats.

I used the following simple benchmark (on Pyboard V1.1 running MicroPython v1.9.2-69-gcc7fece-dirty on 2017-09-12).
```python
import framebuf
from utime import ticks_us, ticks_diff

buf = bytearray(8192)
targets = {
    framebuf.MONO_VLSB : (256, 256, 1, 'MONO_VLSB'),
    framebuf.MONO_HLSB : (256, 256, 1, 'MONO_HLSB'),
    framebuf.MONO_HMSB : (256, 256, 1, 'MONO_HMSB'),
    framebuf.RGB565 : (64, 64, 0xffff, 'RGB565'),
    framebuf.GS4_HMSB : (128, 128, 0xf, 'GS4_HMSB'),
    }

white = 0
for t in targets:
    width, height, black, name = targets[t]
    fb = framebuf.FrameBuffer(buf, width, height, t)
    fb1 = framebuf.FrameBuffer(buf, width, height, t)
    fb.fill(white)
    start = ticks_us()
    fb.fill(black)
    tim = ticks_diff(ticks_us(), start)
    print('{:12s} Fill {:5d}μs  '.format(name, tim), end = '')
    fb1.fill(black)
    start = ticks_us()
    fb.blit(fb1, 0, 0)
    tim = ticks_diff(ticks_us(), start)
    print('Blit {:5d}μs'.format(tim))
```
Outcome with standard build:
```
MONO_VLSB    Fill  3579μs  Blit 45028μs
RGB565       Fill   158μs  Blit  2144μs
GS4_HMSB     Fill   297μs  Blit 10903μs
MONO_HLSB    Fill  7112μs  Blit 43855μs
MONO_HMSB    Fill  7114μs  Blit 43854μs
```
With this patch (adapted for new directory structure):
```
MONO_VLSB    Fill 17630μs  Blit 44253μs
RGB565       Fill   142μs  Blit  2147μs
GS4_HMSB     Fill  4815μs  Blit 11573μs
MONO_HLSB    Fill 17629μs  Blit 40335μs
MONO_HMSB    Fill 17237μs  Blit 39557μs
```

**Comment by deshipu** [comment]
If you look at the numbers for the blit, you will see that there isn't actually much of a difference. The numbers for the fill are worse, because I'm using a generic fill routine for all the modes, that doesn't inline the `set_pixel` function (because it's not known at the compile time which one to use). I suppose you could have separate ones for the separate formats, but I'm not sure it really makes sense to optimize for this particular operation...

Thank you very much for working on this!

**Comment by peterhinch** [comment]
I wondered about the blit results and suspected my benchmark. I just now modified the code to blit from one framebuf to another - the edited code above reflects this. The results are identical to the first version within the limits of experimental error (1 digit of μs). I'm now convinced the blit results are genuine.

As you say the difference resulting from the PR is negligible when it comes to blitting. I'll be interested to see the response of all concerned to the fill results.

**Comment by deshipu** [comment]
I should probably split the fill routine into separate functions for all the modes, and inline the set_pixel calls, then it will be hopefully equally fast (but probably also equally large).
